tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15358,10 +15358,52 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
   return;
 }
 
+void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD) {
+  const std::string &EntryPointName = S.getLangOpts().HLSLEntryFunction;
+
+  // if this is the Entry FD, then add the target profile
+  // shader attribute to the FD and carry on with validation
+  // otherwise, with no shader attribute, just return
+  if (EntryPointName.empty()) {
+    return;
+  }
+  
+  // There's no way we should be adding a shader attribute
+  // to a Function Decl if it doesn't even have an identifer
+  if (!FD->getIdentifier()) {
+    return;
+  }
+
+  // if this FD isn't the entry point, then there's no
+  // shader attribute to work with, so just return
+  if (EntryPointName != FD->getIdentifier()->getName() || 
+    (EntryPointName == FD->getIdentifier()->getName() && !FD->isGlobal())) {
+    return;
+  }
+
+  std::string profile = S.getLangOpts().HLSLProfile;
+  const ShaderModel *SM = hlsl::ShaderModel::GetByName(profile.c_str());
+  const llvm::StringRef fullName = ShaderModel::FullNameFromKind(SM->GetKind());
+
+  // don't add the attribute for an invalid profile, like library
+  if (fullName.empty()) {
+    return;
+  }
+
+  HLSLShaderAttr *pShaderAttr =
+      HLSLShaderAttr::CreateImplicit(S.Context, fullName);
```

Comment:
If we were to define attributes for each of the shader stages in Attr.td (we have some HLSL-specific ones there already), I think we could get away without having to pass strings around. That's something we're trying to do less of.